236. Lowest Common Ancestor of a Binary Tree

```diff
@@ -0,0 +1,50 @@
+/**
+ * Definition for a binary tree node.
+ * struct TreeNode {
+ *     int val;
+ *     TreeNode *left;
+ *     TreeNode *right;
+ *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
+ * };
+ */
+class Solution {
+    public:
+        TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
+            //visit the tree, find out how to get to p and q from root, then find out the common route.
+            vector<int> proute;
+            vector<int> qroute;
+            inorder(root, proute, p);
+            inorder(root, qroute, q);
+            int minl = min(proute.size(), qroute.size());
+            for (int i = 0; i < minl; ++i) {
+                if (proute[i] != qroute[i]) {
+                    return root;
+                }else if (proute[i] == 0) {
+                    root = root->left;
+                }else{
+                    root = root->right;
+                }
+            }
+            return root;
+        }
+        int inorder(TreeNode* root, vector<int>& route, TreeNode* target) 
```

Comment:
I want you to write the recursion function of inorder? Please try to answer following questions when you write the function:

1. What is the meaning of this function?
2. What is the base case.
3. What is the domain?

Reply:
1. I was stupid. I was trying to write preorder as name. This really looks like preorder traversal..
2. Base cases:
    Root == nullptr, return 0 (false);
    Root == target, return 1 (true);
3. I am not quite sure what do you mean by the word domain. Would you brief explain it for me. I am trying to take it as the domains of the mathematical function, and in that case I will give the answer "the whole tree".

Reply:
I will write one for you as an example and you can follow this in the future.
Define function find_route(node, target) as following:
find_route(node, target) = 
1. false, if node = null (Base case)
2. true, if node = target (Base case)
3. f(node->left, target) || f(node->right, target), otherwise.

find_route will return true if there is a path from the node to target, otherwise false.

Note that the domain here is Node X Target. You can take a look at your basic mathematics book, how do they define the domain. Usually, you will see RXR or similar symbols. Here Node is either null or a real node. Target is a real node. So your domain is fairly simple: { {null, real node}, {real node, real node} }. And if you check my domain in the function, you will see I cover all domains. Thus my function is completed. This is how you prove that your solution is correct.
You only write code after you have correct recursion function. And since you have the function, you can write the code easily. I will write it based on my function, and you will see how similar they are.

Speaking of the route, the route is not a parameter of the function, but rather a structure to help record the path. So I didn't put it in the function itself. But it will be written in the code. However, you can write the function more programmatically not mathematically by including route in the function definition.

Reply:
This is really new to me. This is not the way I think when I was writing a program. Programming and math were really something very different for me previously. But this sounds like really a more reliable way of programming and I will try to think in this way in my future code.